### ISSUE TITLE
Add geotype to datatypes

### ISSUE THREAD

**Issue by etiennedi** [creativesoftwarefdn/weaviate#729]
## Todos

* [x] design decisions
  * [x] name of the field
    * current proposals: `geoCoordinate`, `geoLocation`, `geoPoint`
      * my personal favorite being `geoCoordinate`
      * cc @laura-ham @bobvanluijt 
  * [x] design of the where filter
    * @laura-ham suggestions?
* [x] spike out happy path in janusgraph only
  * [x] index creation
  * [x] adding property
  * [x] searching by property within range
* [x] add new data type on import, goal: an import with `geoCoordinates` field succeeds
  * [x] allow in schema creation
  * [x] allow in class instance creation
  * [x] validation?
    * [x] add basic validation
    * [x] refactor `validateSchemaInBody` (it's way too long and extremely difficult to read/extend) 
  * [x] janus graph create vertex
* [x] include on simple read queries
  * [x] Local Get
  * [x] Network Get
* [x] filter by property
  * [x] Local Filters
    * [x] extract filter from graphql
    * [x] set required validations, so that required fields cannot be omitted
    * [x] apply filter in connectors (Janusgraph)
  * ~~Network Filters~~
    * nothing to do here, they use the same code as local filters
* [x] deal with property in `GetMeta` and `Aggregate`
  * proposal for now to simply not support those fields there
  * long-term?
* [x] rename according to latest decisions
  * [x] pluralize name `geoCoordinate` -> `geoCoordinates`
  * [x] restructure where filter
    * [x] `WithinRange` -> `WithinGeoRange`
    * [x] `valueRange` -> `valueGeoRange`
    * [x] wrap `distance` and `geoCoordinates` in separate objects
* [x] Update docs (@laura-ham volunteered to help out here)
* [x] e2e/acceptance test

Original Content below: *for most up-to-date summary see comments below*

## Abstract Examples
*  A location in a 2-dimensional space, `i.e. x=3, y=5`
* Coordinates that point to a location on a world map

## Research Questions
* Are geo types always two-dimensional or can they also be more dimensional?
  * Part 1: Do we have use cases for more than 2d?
  * Part 2: Is there technical support for more than 2d?
* Optimal way to query, e.g. `within 500m coordinates x,y` mixes concepts of metrical distance and coordinates, what API do we want

## Features we'd need
* Import/update geo coordinates
* use geo in where filters (see research question)
* do we want to allow aggregation functions like `Aggregate` or `GetMeta`
  * if so, what should they look like?
  * if so, is there technical support in our current stack?
  * if so, right from the start or later?

cc @bobvanluijt @laura-ham

**Comment by bobvanluijt** [comment]
Please don't forget to add it to the ontology docs :-)

**Comment by laura-ham** [comment]
@etiennedi 

I think we should support at least 3D geo types. A simple example is the altitude of locations on earth (@bobvanluijt could be important for satellite measurements as well). 
But I think we should not adopt the name `geotype` for this datatype, but rather something more abstract/general like `coordinate`. 

> * Optimal way to query, e.g. `within 500m coordinates x,y` mixes concepts of metrical distance and coordinates, what API do we want
To support this, we have to assume that we only applying `geocoordinates`, and not any other coordinate in a 2/3/4/5/..-dimensional space. Also, there are many different geographical coordinate formats around (http://www.avenza.com/help/geographic-imager/5.0/coordinate_formats.htm), so conversion will be needed if we want to support comparisons with meters etc. (assuming the only support is geographical coordinates, not any dimensional space.) In addition, maybe aviation or maritime fields use other coordinate systems..

So I guess the first decision should be made on if we want to support only geographical coordinates (and then, 2d or 3d) or any dimensional space. 
Then, we could start thinking about filters and querying etc.

**Comment by etiennedi** [comment]
Summary of what @laura-ham, @bobvanluijt and I discussed on Slack:

- We accept that there are many use cases for coordinates (different coordinate systems, different dimensions), but specifically want to reduce this problem to “a place on earth”
- At first we will only support `latitude` and `longitude` as the only two parameters (mostly motivated by the fact that a) we see a use case for this in the very short-term and b) our current backends (connectors) don’t support anything else)
- However, in the future we might add Altitude as a third parameter, to effectively make this 2d system a 3d system (however, still limited to places on (or above/below) earth
- Our main goal is to able to query `within x kilometers of point<lat,long>` -> We have not talked about how this filter would look like, but would do so next.
- The new datatype would be entered as a map like `nameOfNewType: {"latitude": 0.12345, "longitude": 0.9875}` whereas the numbers represent the *decimal* format. We have not talked about what this prop should be called, but would do so next.

**Comment by etiennedi** [comment]
Janusgraph Research Results

add a field called location that is a Geoshape and add it to the existing (elasticsearch-based) index:
```
graph.tx().rollback()
mgmt = graph.openManagement()
  newProp = mgmt.makePropertyKey("location").dataType(Geoshape.class).cardinality(Cardinality.SINGLE).make()
  existingIndex = mgmt.getGraphIndex("search")
  mgmt.addIndexKey(existingIndex, newProp)
mgmt.commit()
```

Add a vertex with the coordinates of Amsterdam:
```
g.addV().property("location", Geoshape.point(52.366667, 4.9)) // Amsterdam
```

Try to find it based on the coordinates of Düsseldorf:
```
g.V().has("location", geoWithin(Geoshape.circle(51.233333, 6.783333, 100)))
// -> no results

g.V().has("location", geoWithin(Geoshape.circle(51.233333, 6.783333, 300)))
// found the vertex we created above
```

**Comment by bobvanluijt** [comment]
Something on a meta-level (not necessarily related to Geoloc) we should make a distinction between `types` and `formats` (as you can already see [here](https://github.com/creativesoftwarefdn/weaviate/blob/develop/docs/en/use/ontology-schema.md#property-datatypes)). Especially from a connector PoV this is important, it might be the case that a non-JanusGraph connector is unable to deal with functionality like `geoWithin()`. There is still value though because it informs an end-user about what a type might contain. 

Example (note the `@dataType`):

```json
{
    "@context": "http://semi.network",
    "classes": [
        {
            "class": "Car",
            "description": "This is a car",
            "keywords": [
                {
                    "keyword": "Vehicle",
                    "weight": 1
                }
            ],
            "properties": [
                {
                    "@dataType": [
                        "geoLocation"
                    ],
                    "description": "location of the car",
                    "keywords": [],
                    "name": "location"
                }
            ]
        }
    ]
}
```

The more descriptive it is, the better the end-user will be able to use it (for example, @laura-ham's point about using `coordinate` would be too ambiguous. However, it is simple to add both).

**Use case example;**
- User requests meta information from meta-endpoint.
- The user learns that `Car.location` == `geoLocation`
- The user can now use the coordinates to plot these -for example- on a map.

**Completely different example;**
- `dataType==float64` and `dataFormat==temprature`.
- This is minimal but helpful for end-users

This means a few things;
- Without the connector functionality, this is already valuable.
- We now call it `@dataType` this should actually be: `@dataFormat` where a `string` is also seen as a format.
- @etiennedi can add functionality when needed (while keeping it connector agnostic).

**Comment by etiennedi** [comment]
@laura-ham I just realized that I was sometimes using the name `geoCoordinate` (singular) and sometimes `geoCoordinates`. I just changed it to make sure it's unique throughout, but I'm not sure yet which option is better.

Is there such a thing as a single coordinate? Or would the combination of latitude and longitude both be a coordinate, i.e. we'd always have a pair of coordinates? If so, should the type be called `geoCoordinates`?

**Comment by etiennedi** [comment]
@bobvanluijt would mind creating a separate issue for https://github.com/creativesoftwarefdn/weaviate/issues/729#issuecomment-467774557. I think I agree with what you're saying, but I'd like to tackle this independently of this issue. Thansk

**Comment by etiennedi** [comment]
@laura-ham I'll most likely get to the point today where I have to build the where filter for a "range search". Do you have any spontaneous design idea? If not, I can also just build something that works, and then we can make it nicer whenever you have some time.

**Comment by laura-ham** [comment]
> @laura-ham I just realized that I was sometimes using the name `geoCoordinate` (singular) and sometimes `geoCoordinates`. I just changed it to make sure it's unique throughout, but I'm not sure yet which option is better.
> 
> Is there such a thing as a single coordinate? Or would the combination of latitude and longitude both be a coordinate, i.e. we'd always have a pair of coordinates? If so, should the type be called `geoCoordinates`?

I think there is only a plural `Coordinates` in our case, we always have at least a latitude and longitude. (we were not the first ones wondering this https://www.quora.com/Are-coordinates-plural-or-singular) However, I see Janusgraph uses the singular form `GeoCoordinate`..



> @laura-ham I'll most likely get to the point today where I have to build the where filter for a "range search". Do you have any spontaneous design idea? If not, I can also just build something that works, and then we can make it nicer whenever you have some time.

TLDR; see last piece of code and explanation below :)

I think we should keep as close to the original design as possible, athough we indeed have to make some additions to make it work for coordinates as well. Until now, `value<...>` is based on the `dataType`s, and points to a single value to compare the data with. Here, we however need more input; we need the coordinates and the distance. I am wondering if it would be a problem for the user to make this a dict instead of a single input. For example:

```graphql
{
  Local {
    Get {
      Things {
        City(where: {
          operator: WithinGeoRange,
          valueGeoRange: {geoCoordinates: {latitude: 0.5, longitude: 0.5}, distance: 2.0},
          path: ["geolocation"]
        }) {
          name
          geolocation {
            latitude
            longitude 
          }
        }
      }
    }
  }
}
```

Do we need to introduce new operators like `Within` and `Outside`? Or could we use the existing operators `LessThan`, `LessThanEqual`, `Equal` (unlikely to be used, but still), `GreaterThan` and `GreaterThanEqual`? So:

```graphql
{
  Local {
    Get {
      Things {
        City(where: {
          operator: LessThanEqual,
          valueGeoRange: {geoCoordinates: {latitude: 0.5, longitude: 0.5}, distance: 2.0},
          path: ["geolocation"]
        }) {
          name
          geolocation {
            latitude
            longitude 
          }
        }
      }
    }
  }
}
```

I would go for `Within` and `Outside` now that I think of it... What do you think?

An alternative might be: 
```graphql
{
  Local {
    Get {
      Things {
        City(where: {
          operator: LessThanEqual,
          valueGeoDistance: {geoCoordinates: {latitude: 0.5, longitude: 0.5}, distance: 2.0},
          path: ["geolocation"]
        }) {
          name
          geolocation {
            latitude
            longitude 
          }
        }
      }
    }
  }
}
```

Then what came to my mind is what if you want to find locations in a range with a from and to distance specification:

```graphql
{
  Local {
    Get {
      Things {
        City(where: {
          operator: WithinGeoRange,
          valueGeoRange: {geoCoordinates: {latitude: 0.5, longitude: 0.5}, distance: {from:2.0, to:3.0}},
          path: ["geolocation"]
        }) {
          name
          geolocation {
            latitude
            longitude 
          }
        }
      }
    }
  }
}
```
And if you only want to specify one limit:
```graphql
{
  Local {
    Get {
      Things {
        City(where: {
          operator: WithinGeoRange,
          valueGeoRange: {geoCoordinates: {latitude: 0.5, longitude: 0.5}, distance: {to:2.0}},
          path: ["geolocation"]
        }) {
          name
          geolocation {
            latitude
            longitude 
          }
        }
      }
    }
  }
}
```

Sorry for posting my whole thought process here. I think the last design would be the best. We need to introduce two new operators: `WithinGeoRange` and `OutsideGeoRange` (I think mentioning `Geo` in the name is good because this filter will only work on GeoCoordinates, not on other datatypes). 

(I am wondering, if it would be possible to query the `latitude` and `longitude` separately, since these are parts of a single property datatype?)

We would have to specify in the docs that the distance should always be in kilometers (and also the correct format of coordinates of course).

**Comment by etiennedi** [comment]
> Sorry for posting my whole thought process here. I think the last design would be the best. We need to introduce two new operators: `WithinGeoRange` and `OutsideGeoRange` (I think mentioning `Geo` in the name is good because this filter will only work on GeoCoordinates, not on other datatypes).

Agreed. I'll update it.

However, I'll only implement the case where only `distance: { to }` is set for now. And outscope the `from` part. This is because we're on a tight timeline to v1 and I need to cut corners whereever possible. We can then introduce the additional `from` field in a later minor version. (cc @bobvanluijt) 

Possibly the same for `OutsideGeoRange`, then again, that's probably just adding a `not()` operator.

> (I am wondering, if it would be possible to query the `latitude` and `longitude` separately, since these are parts of a single property datatype?)
Not with janus, there it's a single entity (`Geoshape.Point` that requires both)

> We would have to specify in the docs that the distance should always be in kilometers (and also the correct format of coordinates of course).

Yes, I've always started that for the graphQL description fields, but I haven't touched the docs yet until we have finalized the design a bit. Since I could use every help possible, do you have some time to take that up by any chance?

**Comment by laura-ham** [comment]
> However, I'll only implement the case where only `distance: { to }` is set for now. And outscope the `from` part. This is because we're on a tight timeline to v1 and I need to cut corners whereever possible. We can then introduce the additional `from` field in a later minor version. (cc @bobvanluijt)

Yes, makes sense. 

> Possibly the same for `OutsideGeoRange`, then again, that's probably just adding a `not()` operator.

True, let's remove `OutsideGeoRange`, because this is indeed the same as applying the `not` operator.

> > (I am wondering, if it would be possible to query the `latitude` and `longitude` separately, since these are parts of a single property datatype?)
> Not with janus, there it's a single entity (`Geoshape.Point` that requires both)

So this cannot be requested as separate properties as we do in the query above?

> Yes, I've always started that for the graphQL description fields, but I haven't touched the docs yet until we have finalized the design a bit. Since I could use every help possible, do you have some time to take that up by any chance?

I could do the docs, but not before tomorrow. Do you also mean https://github.com/creativesoftwarefdn/weaviate/blob/develop/docs/en/use/ontology-schema.md#property-datatypes in addition to GraphQL docs?

**Comment by etiennedi** [comment]
> So this cannot be requested as separate properties as we do in the query above?

From a db perspective (in the janus case) we retrieve the pair as an atomic unit. However, in the graphql resolver we only show what the user requested. So if they only want to show the `Latitude`, but not the `Longitude` that's fine. 

However, if they try to perform transaction on it (such as a within-range search) both fields have to be provided. This is also how I've currently implemented it on the first iteration of the where filter I've just build. All three fields (lat, lon, dist) are required.

> I could do the docs, but not before tomorrow.

That would be absolutely fine. Also later would be fine. In fact this might be good, because then you could actually verify that the way I've implemented matches what you want to document. 

Yes, so far I've only filled the `description` fields in the graphql API itself. This might also make sense to review, but it's only 3 sentences so far. I haven't touched any markdown files yet. :)

Thanks a lot

**Comment by laura-ham** [comment]
> > So this cannot be requested as separate properties as we do in the query above?
> 
> From a db perspective (in the janus case) we retrieve the pair as an atomic unit. However, in the graphql resolver we only show what the user requested. So if they only want to show the `Latitude`, but not the `Longitude` that's fine.

Okay thanks, I get that. But it is still not clear how the user then can request to get the value for longitude and latitude back as separate property values, while it is in the database formatted into one property.

**Comment by etiennedi** [comment]
Not sure I understand the last part.

In the db the type of the field that I call `geolocation` above is a `Geoshape.Point`. This Java class in turn has properties for the coordinates. I think it's an array of two floats whereas element 0 is lat and element 1 is long. This is at the same level where the prop `name` would be the java class `String` which doesn't have any properties, but is just the string itself.

In the graphql resolver I can then transform this array into the respective fields. This is currently working already:

![image](https://user-images.githubusercontent.com/8974479/53565746-89698480-3b5a-11e9-844a-3475a78b6b6d.png)

Does this help? Or am I completely misunderstanding the problem?

**Comment by laura-ham** [comment]
Okay, so by default there is a `latitude` and `longitude` field in every property with the @datatype `GeoCoordinates`?

**Comment by laura-ham** [comment]
Okay I completely understand it now :) thanks for the clear explanation!

**Comment by etiennedi** [comment]
@laura-ham 
Spontaneous thought: Would `distance: { min max }` be easier to understand than `distance { from to }`. I find from to a bit ambiguous in the setting of coordinates and maps, because it reminds me of taking a measurement from a specific point to another. Wondering what you think about this.

**Comment by laura-ham** [comment]
If we talk about `distance` then I agree that `min` and `max` would be better options. I initially copied the `from` and `to` from https://www.compose.com/articles/geofile-elasticsearch-geo-queries-2/, using them for `geo_distance_range`.

**Comment by etiennedi** [comment]
Cool, then I'll reflect that as well. :)

**Comment by etiennedi** [comment]
This is is the flow mentioned in https://github.com/creativesoftwarefdn/weaviate/issues/729#issuecomment-468314128 with the updated names:

```json
{
    "name": "geolocation",
    "@dataType": [
	"geoCoordinates"
    ],
    "description": "The city's longitude and latitude",
    "keywords": [
      { "keyword": "location", "weight": 1.0 }
    ]
}
```


and on import we have to specify something like this:


```json
"schema": {
    "name": "Amsterdam",
    "population": "1800000",
    "isCapital": true,
    "geolocation": {
	"latitude": 52.366667,
	"longitude": 4.9
    }
}
```

**Comment by bobvanluijt** [comment]
@laura-ham because I'm working on the docs I've already tackled this.
